### PR TITLE
discriminator objective

### PR DESCRIPTION
**Pull request by coopie**
Summary:

### Adds a discriminator based off the one used in the SpeechPainter paper
the adversarial training works as below:
* The discriminator only sees a 30-wide window of a spectrogram, so first we slice the reference and predicated spectrograms into 30-width chunks
* A random sample of `batch_size` slices are chosen from the real and generated spectrogram slices. Note that the same slice indices are needed for both real and generated spectrogram slices for the training to work properly (i.e. if the batch is constructed from indices 1, 10, 99, 5, for the real spectrograms, the same indices must be used to create the batch for the generated samples)
* The discriminator is trained with a hinge loss on the logits of the real and generated spectrogram predictions
* The generator (i.e. inpainter model) has another objective added to it - matching the activations of a real and generated spectrogram slice inside the discriminator (also called feature matching loss)

### Adds truncating of input spectrograms if they are longer than a given number of windows
This guarantees that we aren't going to have an occasional OOM error when training the model

### Fixes a bug in `regenerate_audio` that improves the quality of recordings considerably
Samples can be found here [https://docs.google.com/presentation/d/1mf_UVpA-2iadDHwflUB3A7YeQB8VpZtN2xVDgO3W2ng/edit?usp=sharing

### Updates to the readme on how to train the inpainter and vocoder on combined data
It's not one-click but it should be fairly quick for people to duplicate the work

### CODE REVIEW

**Comment by pawel-polyai** [comment]
Once polished, can you add the exact config/recipe to run the model with discriminator loss with the results saved add GCP?

File: README_INPAINTING.md [hunk 1]
```diff
@@ -1,53 +1,69 @@
...
+python -m venv .env
+source .env/bin/activate
 ```
 
+Then install the dependencies needed to run NeMo
 
-(optional) download cached computed data
-download from [here](https://drive.google.com/file/d/1lyZHm2KdRoBJgW2XuCV9q_EaXquw9wgc/view?usp=sharing)
-
-then unzip and place as `data/LJSpeech-1.1/data_cache` this will speed up the first epoch on the machine greatly
```
Comment (pawel-polyai) [inline README_INPAINTING.md:26]: Inside the inpainting folder?

File: README_INPAINTING.md [hunk 2]
```diff
@@ -1,53 +1,69 @@
...
+### Training (HiFiGan)
+
+Training the HiFIGAN vocoder is very similar to the inpainter model. to train it on all data, run:
 
-You will have to create the smaller datasets which can be quickly done by running:
 ```bash
-cat data/LJSpeech-1.1/test_manifest.json | head -n 32 > data/LJSpeech-1.1/test_manifest_tiny.json
-cat data/LJSpeech-1.1/val_manifest.json | head -n 32 > data/LJSpeech-1.1/val_manifest_tiny.json
+python examples/tts/hifigan.py train_dataset=data/ltts_vctk_nicky/combined_train.json validation_datasets=data/ltts_vctk_nicky/combined_val.json
+
```
Comment (pawel-polyai) [inline README_INPAINTING.md:68]: Can you add explicit function calls in examples/tts/conf and the latest model versions in the gcp of the inpainter, aligner and vocoders?

File: generate_eval_audios.py [hunk 3]
```diff
@@ -20,19 +20,87 @@
 
 model = InpainterModel.load_from_checkpoint(checkpoint_path)
 
+vocoder = HifiGanModel.from_pretrained(model_name="nvidia/tts_hifigan")
```
Comment (pawel-polyai) [inline generate_eval_audios.py:23]: This is hard to read. Can we create a separate PR that will use the latest model trained with Nicky data where we will produce samples for both full sentence generation with inpainting and inpainting with concatenation to the original recordings with samples downsized to 8k?

File: README_INPAINTING.md (hunk 2 above)
Comment (coopie) [reply README_INPAINTING.md:68]: not sure what you mean by explicit function calls

File: generate_eval_audios.py [hunk 4]
```diff
@@ -55,33 +34,123 @@ def get_all_replacements(text):
 
 expected_sample_rate = 22050
 
-vocoder = HifiGanModel.from_pretrained(model_name="nvidia/tts_hifigan")
 
-
-def apply_repalcements(file_path, replacement_phrase, dest):
+def apply_repalcements(file_path, replacement_phrase, dest, inpainter, vocoder):
```
Comment (pawel-polyai) [inline generate_eval_audios.py:38]:
```suggestion
def apply_replacements(file_path, replacement_phrase, dest, inpainter, vocoder):
```

File: generate_eval_audios.py [hunk 5]
```diff
@@ -1,37 +1,16 @@
 import soundfile as sf
+import hashlib
```
Comment (pawel-polyai) [inline generate_eval_audios.py:2]: Can you add the header with the exemplary run?

File: examples/tts/conf/inpainting.yaml [hunk 6]
```diff
@@ -150,7 +150,7 @@ model:
       drop_last: false
       shuffle: false
       batch_size: 16
-      num_workers: 1
+      num_workers: 4
```
Comment (pawel-polyai) [inline examples/tts/conf/inpainting.yaml:153]: Can you add the header with the command that trained the latest model and its localization on the GCP?

File: examples/tts/conf/fastpitch_align_44100.yaml [hunk 7]
```diff
@@ -247,4 +247,3 @@ exp_manager:
     monitor: val_loss
```
Comment (pawel-polyai) [inline examples/tts/conf/fastpitch_align_44100.yaml:247]: Can you add the header with the command that trained the latest model and its localization on the GCP?

**Review by pawel-polyai — COMMENTED** [review]
Can you also add the vocoder run and the path to the GCP?

File: README_INPAINTING.md (hunk 2 above)
Comment (pawel-polyai) [reply README_INPAINTING.md:68]: Sorry, I mean saving the exact command that you used to train the model.

**Review by pawel-polyai — APPROVED** [review]
Lgtm
